README.rst

```diff
@@ -19,59 +19,117 @@ Simply feed in a raw string of text (or HTML), and receive a list of structured
 
-Options
-=======
-:code:`get_citations()`, the main executable function, takes several parameters.
+::
 
-1. :code:`do_post_citation` ==> bool; whether additional, post-citation information should be extracted (e.g., the court, year, and/or date range of the citation)
-2. :code:`do_defendant` ==> bool; whether the pre-citation defendant (and possibily plaintiff) reference should be extracted
-3. :code:`disambiguate` ==> bool; whether each citation's (possibly ambiguous) reporter should be resolved to its (unambiguous) form
-4. :code:`clean` ==> tuple; a tuple of cleaning steps to undertake before parsing; options include :code:`whitespace` (remove extraneous whitespace [default]), :code:`underscores` (remove extraneous underscores), and :code:`html` (remove non-visible HTML content)
-5. :code:`tokenizer` ==> Tokenizer; an instance of a Tokenizer object (see "Tokenizers" below)
+    from eyecite import get_citations, annotate, clean_text
 
-Some notes
-----------
-Some things to keep in mind are:
+    source_text = '<p>bob lissner v. <i>test   1 U.S.</i> 12,   347-348 (4th Cir. 1982)</p>'
+    plain_text = clean_text(source_text, ['html', 'whitespace'])
+    citations = get_citations(plain_text)
+    linked_text = annotate(plain_text, [[c.span(), "<a>", "</a>"] for c in citations], source_text=source_text)
 
-1. This project depends on information made available in two other Free Law Project packages, `reporters-db <https://github.com/freelawproject/reporters-db>`_ and `courts-db <https://github.com/freelawproject/courts-db>`_.
-2. This package performs no matching or resolution action. In other words, it is up to the user to decide what to do with the "short form," "supra," "id.," and "ibid." citations that this tool extracts. In theory, these citations are all references to "full" citations also mentioned in the text -- and are therefore in principle resolvable to those citations -- but this task is beyond the scope of this parsing package. See `here <https://github.com/freelawproject/courtlistener/tree/master/cl/citations>`_ for an example of how Courtlistener implements this package and handles this problem.
+    returns:
+    '<p>bob lissner v. <i>test   <a>1 U.S.</a></i><a> 12</a>,   347-348 (4th Cir. 1982)</p>'
+
+The above example extracts citations from :code:`plain_text` and applies them to
+:code:`source_text`, using a diffing algorithm to insert annotations in the correct locations
+in the original text.
+
+Note that the annotations are wrapped around each spot where the cleaning steps removed text
+from the citation: "<a>1 U.S.</a></i><a> 12</a>". This ensures that :code:`annotate` will
+emit valid HTML. If you don't want that, use :code:`wrap_elisions=False`:
+
+::
+
+    linked_text = annotate(plain_text, [[c, "<a>", "</a>"] for c in citations], source_text=source_text, wrap_elisions=False)
+
+    returns:
+    '<p>bob lissner v. <i>test   <a>1 U.S.</i> 12</a>,   347-348 (4th Cir. 1982)</p>'
```

Comment:
IMO it would be preferable for the default behavior to be to _not_ wrap elisions (i.e., to annotate the entire string as one). But, as you point out, the risk of doing this is that sometimes broken HTML will be emitted (if the original text has intervening HTML tags, as in this example). To overcome this problem in CL I created a function `is_balanced_html()` (https://github.com/freelawproject/courtlistener/blob/master/cl/citations/utils.py#L47) so we could first check whether the text-to-be-annotated indeed contained such intervening tags (i.e., was "unbalanced" HTML). Could we do something similar here?

This would be my ideal behavior, I think:
- `annotate()` tries to wrap the entire string as one (no flag needed)
- If `annotate()` cannot do so safely (i.e., the text-to-be-annotated fails the `is_balanced_html()` check), then a new flag determines how it proceeds, in one of three ways:
  - It does the full wrap anyway, knowing that broken HTML will be emitted
  - It safely wraps around the individual elisions instead (the current behavior under `wrap_elisions=True`)
  - It does no annotation at all

Thoughts?

Reply:
I'll take a stab at this -- it's nice to not wrap text replacements like smart quotes. I think the default wants to be no check, since "this is HTML with spans and I should do an expensive parsing step for each string" is something you should have to opt into.

Reply:
OTOH, I think if it's the only way to get good results, that kind of expensive parsing might just be the cost of doing business, and having defaults with good (albeit slow) results might just be OK.

Reply:
My latest update just inserts annotations at the edges of the citation by default, but accepts `unbalanced_tags="wrap"` to wrap HTML tags in unbalanced tags are detected, and `unbalanced_tags="skip"` to skip those annotations.

One choice I made here is to only wrap HTML tags, using a new `wrap_html_tags` helper, and not wrap all elisions like extra spaces. This is prettier, and also I thought it would be a weird behavior if spaces were wrapped or not based on whether or not the citation also contained an unbalanced tag.

A second choice I made was to assume straightforward HTML tags in `wrap_html_tags` that can be matched with `r"<[^>]+>"`. This isn't correct for all HTML, but wrapping HTML text spans is a large problem on arbitrary HTML, and I documented that as out of scope in the readme -- we're assuming your HTML is properly escaped (you could process it through lxml.html if not), and that your citations don't include anything super weird like the start of an HTML comment or a pre tag (I think this would only happen with deliberately broken input). If malicious input is a concern one should just insert self-closing markers and do the highlighting with a heavier library, maybe in javascript. (Note this choice doesn't depend on the previous choice -- you could still make broken HTML from malicious input if we were wrapping all elisions instead of just tags.)

> I think if it's the only way to get good results, that kind of expensive parsing might just be the cost of doing business, and having defaults with good (albeit slow) results might just be OK.

For this version I stuck with the default of not checking for unbalanced tags, though it's not a huge deal to me either way. I agree that wrapping is a good default _if_ your source_text is (a) HTML (b) that is reasonably well-formed but (c) has potential markup in the citations. I just think that's specific enough that you should have to think about it and make sure it applies to you; it shouldn't be an implicit contract of the default `annotate()` call.

This isn't a performance issue since we can short-circuit most strings by checking for angle brackets, it's more just a heavy amount of code to run that might do stuff that doesn't make sense outside our particular situation.